tools/ctl/ctl.go

```diff
@@ -229,34 +271,181 @@ func makeHTTPRequest(ctx context.Context, u *url.URL, msg schema.Message) *http.
 	return req
 }
 
+type WorkerConfig struct {
+	client   *http.Client
+	url      *url.URL
+	limiters map[string]<-chan time.Time
+	run      string
+}
+
+type AttestWorker struct {
+	WorkerConfig
+}
+
+func (w *AttestWorker) processOne(ctx context.Context, p benchmark.Package, out chan schema.Verdict) {
+	for _, v := range p.Versions {
+		<-w.limiters[p.Ecosystem]
+		resp, err := w.client.Do(makeHTTPRequest(ctx, w.url.JoinPath("rebuild"), schema.RebuildPackageRequest{
+			Ecosystem: rebuild.Ecosystem(p.Ecosystem),
+			Package:   p.Name,
+			Version:   v,
+			ID:        w.run,
+		}))
+		var errMsg string
+		if err != nil {
+			errMsg = errors.Wrap(err, "sending request").Error()
+		}
+		if resp.StatusCode != 200 {
+			errMsg = errors.Wrapf(errors.New(resp.Status), "sending request").Error()
+		}
+		var verdict schema.Verdict
+		if errMsg != "" {
+			verdict = schema.Verdict{
+				Target: rebuild.Target{
+					Ecosystem: rebuild.Ecosystem(p.Ecosystem),
+					Package:   p.Name,
+					Version:   v,
+				},
+				Message: errMsg,
+			}
+		} else {
+			// TODO: Once the attestation endpoint returns verdict objects,
+			// support that here.
+			verdict = schema.Verdict{
+				Target: rebuild.Target{
+					Ecosystem: rebuild.Ecosystem(p.Ecosystem),
+					Package:   p.Name,
+					Version:   v,
+				},
+				Message: "",
+			}
+		}
+		out <- verdict
+	}
+}
+
+func (w *AttestWorker) Run(ctx context.Context, bar *pb.ProgressBar, out chan schema.Verdict, in chan benchmark.Package) {
+	for p := range in {
+		w.processOne(ctx, p, out)
+		bar.Increment()
+	}
+}
+
+type SmoketestWorker struct {
+	WorkerConfig
+	warmup bool
+}
+
+func (w *SmoketestWorker) processOne(ctx context.Context, p benchmark.Package, out chan schema.Verdict) {
+	<-w.limiters[p.Ecosystem]
+	resp, err := w.client.Do(makeHTTPRequest(ctx, w.url.JoinPath("smoketest"), schema.SmoketestRequest{
+		Ecosystem: rebuild.Ecosystem(p.Ecosystem),
+		Package:   p.Name,
+		Versions:  p.Versions,
+		ID:        w.run,
+	}))
+	var errMsg string
+	if err != nil {
+		errMsg = errors.Wrap(err, "sending request").Error()
+	}
+	if resp.StatusCode != 200 {
+		errMsg = errors.Wrapf(errors.New(resp.Status), "sending request").Error()
+	}
+	if errMsg != "" {
+		for _, v := range p.Versions {
+			out <- schema.Verdict{
+				Target: rebuild.Target{
+					Ecosystem: rebuild.Ecosystem(p.Ecosystem),
+					Package:   p.Name,
+					Version:   v,
+				},
+				Message: errMsg,
+			}
+		}
+	} else {
+		d := json.NewDecoder(resp.Body)
+		var r schema.SmoketestResponse
+		if err := d.Decode(&r); err != nil {
+			log.Fatalf("Failed to decode smoketest response: %v", err)
+		}
+		for _, v := range r.Verdicts {
+			out <- v
+		}
+	}
+}
+
+func (w *SmoketestWorker) Run(ctx context.Context, bar *pb.ProgressBar, out chan schema.Verdict, in chan benchmark.Package) {
+	if w.warmup {
```

Comment:
I'd probably move this to a separate Init() method so then Run() can be replaced with the processOne impl (i.e. made scalar).

Reply:
Done.